File: frontend/src/services/whitelabelService.js
/**
 * Whitelabel API Service
 * Handles all white-label configuration API calls
 */

const API_BASE = process.env.REACT_APP_API_URL || 'http://localhost:8000';

/**
 * Get authorization headers
 */
const getHeaders = () => {
  const headers = {
    'Content-Type': 'application/json',
  };

  const token = localStorage.getItem('authToken') || sessionStorage.getItem('authToken');
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  return headers;
};

/**
 * Get headers for multipart/form-data (file uploads)
 */
const getMultipartHeaders = () => {
  const headers = {};

  const token = localStorage.getItem('authToken') || sessionStorage.getItem('authToken');
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  // Don't set Content-Type for multipart - browser will set it with boundary
  return headers;
};

/**
 * Create a new agency
 */
export const createAgency = async (name, description = null) => {
  try {
    const response = await fetch(`${API_BASE}/api/whitelabel/agency`, {
      method: 'POST',
      headers: getHeaders(),
      body: JSON.stringify({ name, description })
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to create agency');
    }

    return await response.json();
  } catch (error) {
    console.error('Error creating agency:', error);
    throw error;
  }
};

/**
 * Get white-label settings for an agency
 */
export const getWhitelabelSettings = async (agencyId) => {
  try {
    const response = await fetch(
      `${API_BASE}/api/whitelabel/settings?agency_id=${agencyId}`,
      {
        method: 'GET',
        headers: getHeaders()
      }
    );

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch settings');
    }

    return await response.json();
  } catch (error) {
    console.error('Error fetching whitelabel settings:', error);
    throw error;
  }
};

/**
 * Update white-label settings for an agency
 */
export const updateWhitelabelSettings = async (agencyId, settings) => {
  try {
    const response = await fetch(
      `${API_BASE}/api/whitelabel/settings?agency_id=${agencyId}`,
      {
        method: 'PUT',
        headers: getHeaders(),
        body: JSON.stringify(settings)
      }
    );

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to update settings');
    }

    return await response.json();
  } catch (error) {
    console.error('Error updating whitelabel settings:', error);
    throw error;
  }
};

/**
 * Upload a logo to Supabase Storage
 */
export const uploadLogo = async (agencyId, file) => {
  try {
    const formData = new FormData();
    formData.append('agency_id', agencyId);
    formData.append('logo', file);

    const response = await fetch(`${API_BASE}/api/whitelabel/logo/upload`, {
      method: 'POST',
      headers: getMultipartHeaders(),
      body: formData
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to upload logo');
    }

    return await response.json();
  } catch (error) {
    console.error('Error uploading logo:', error);
    throw error;
  }
};

/**
 * Verify custom domain
 */
export const verifyDomain = async (agencyId, domain, verificationToken = null) => {
  try {
    const response = await fetch(
      `${API_BASE}/api/whitelabel/domain/verify?agency_id=${agencyId}`,
      {
        method: 'POST',
        headers: getHeaders(),
        body: JSON.stringify({
          domain,
          verification_token: verificationToken
        })
      }
    );

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to verify domain');
    }

    return await response.json();
  } catch (error) {
    console.error('Error verifying domain:', error);
    throw error;
  }
};

export default {
  createAgency,
  getWhitelabelSettings,
  updateWhitelabelSettings,
  uploadLogo,
  verifyDomain
};

File: backend/app/services/whitelabel_service.py
"""
WhitelabelService - Handles white-label configuration and asset storage.
Uses Supabase Storage for logo/favicon uploads with CDN delivery.
"""
import os
import uuid
import logging
from typing import Dict, Optional, BinaryIO
from pathlib import Path
from supabase import create_client, Client
from sqlalchemy.orm import Session

from app.models.agency import Agency
from app.core.config import settings

logger = logging.getLogger(__name__)

class WhitelabelService:
    """Service for managing white-label settings and asset uploads"""

    def __init__(self):
        """Initialize Supabase client for storage operations"""
        self.supabase_url = settings.SUPABASE_URL or settings.REACT_APP_SUPABASE_URL
        self.supabase_key = settings.SUPABASE_SERVICE_ROLE_KEY

        if not self.supabase_url or not self.supabase_key:
            logger.warning("Supabase credentials not configured - storage features will be disabled")
            self.supabase: Optional[Client] = None
        else:
            self.supabase: Client = create_client(self.supabase_url, self.supabase_key)

        # Storage bucket name for whitelabel assets
        self.bucket_name = "whitelabel-assets"

    def _ensure_bucket_exists(self):
        """Ensure the whitelabel assets bucket exists"""
        if not self.supabase:
            raise Exception("Supabase client not initialized")

        try:
            # Try to get bucket
            buckets = self.supabase.storage.list_buckets()
            bucket_names = [b['name'] for b in buckets]

            if self.bucket_name not in bucket_names:
                # Create bucket with public access
                self.supabase.storage.create_bucket(
                    self.bucket_name,
                    options={"public": True}
                )
                logger.info(f"Created Supabase storage bucket: {self.bucket_name}")
        except Exception as e:
            logger.error(f"Error ensuring bucket exists: {e}")
            # Bucket might already exist, continue anyway

    def upload_logo(
        self,
        agency_id: str,
        file_content: bytes,
        filename: str,
        content_type: str = "image/png"
    ) -> Dict[str, str]:
        """
        Upload a logo to Supabase Storage

        Args:
            agency_id: UUID of the agency
            file_content: Binary content of the file
            filename: Original filename
            content_type: MIME type of the file

        Returns:
            Dict with 'url' and 'path' keys
        """
        if not self.supabase:
            raise Exception("Supabase storage not configured")

        try:
            self._ensure_bucket_exists()

            # Generate unique filename
            file_ext = Path(filename).suffix
            unique_filename = f"{agency_id}/logo-{uuid.uuid4()}{file_ext}"

            # Upload to Supabase Storage
            response = self.supabase.storage.from_(self.bucket_name).upload(
                path=unique_filename,
                file=file_content,
                file_options={
                    "content-type": content_type,
                    "cache-control": "3600",
                    "upsert": "true"
                }
            )

            # Get public URL
            public_url = self.supabase.storage.from_(self.bucket_name).get_public_url(unique_filename)

            logger.info(f"Uploaded logo for agency {agency_id}: {public_url}")

            return {
                "url": public_url,
                "path": unique_filename,
                "bucket": self.bucket_name
            }

        except Exception as e:
            logger.error(f"Error uploading logo: {e}")
            raise Exception(f"Failed to upload logo: {str(e)}")

    def delete_logo(self, file_path: str) -> bool:
        """
        Delete a logo from Supabase Storage

        Args:
            file_path: Path to the file in storage (e.g., "agency-id/logo-uuid.png")

        Returns:
            True if deleted successfully
        """
        if not self.supabase:
            raise Exception("Supabase storage not configured")

        try:
            self.supabase.storage.from_(self.bucket_name).remove([file_path])
            logger.info(f"Deleted logo: {file_path}")
            return True
        except Exception as e:
            logger.error(f"Error deleting logo: {e}")
            return False

    def get_agency_settings(
        self,
        db: Session,
        agency_id: str,
        user_id: str
    ) -> Dict:
        """
        Get white-label settings for an agency

        Args:
            db: Database session
            agency_id: UUID of the agency
            user_id: UUID of the requesting user (for authorization)

        Returns:
            White-label settings dict
        """
        # Get agency from database
        agency = db.query(Agency).filter(Agency.id == agency_id).first()

        if not agency:
            raise Exception(f"Agency not found: {agency_id}")

        # Verify user has access (owner or team member)
        if str(agency.owner_user_id) != user_id:
            # TODO: Check if user is team member
            raise Exception("Unauthorized: User does not have access to this agency")

        return agency.whitelabel_settings

    def update_agency_settings(
        self,
        db: Session,
        agency_id: str,
        user_id: str,
        settings: Dict
    ) -> Dict:
        """
        Update white-label settings for an agency

        Args:
            db: Database session
            agency_id: UUID of the agency
            user_id: UUID of the requesting user
            settings: New settings to apply

        Returns:
            Updated settings
        """
        # Get agency from database
        agency = db.query(Agency).filter(Agency.id == agency_id).first()

        if not agency:
            raise Exception(f"Agency not found: {agency_id}")

        # Verify user has access
        if str(agency.owner_user_id) != user_id:
            raise Exception("Unauthorized: User does not have access to this agency")

        # Update settings
        agency.update_whitelabel_settings(settings)
        db.commit()
        db.refresh(agency)

        logger.info(f"Updated white-label settings for agency {agency_id}")

        return agency.whitelabel_settings

    def create_agency_for_user(
        self,
        db: Session,
        user_id: str,
        name: str,
        description: Optional[str] = None
    ) -> Agency:
        """
        Create a new agency for a user

        Args:
            db: Database session
            user_id: UUID of the user creating the agency
            name: Agency name
            description: Optional agency description

        Returns:
            Created Agency model
        """
        # Check if user already has an agency
        existing = db.query(Agency).filter(Agency.owner_user_id == user_id).first()
        if existing:
            logger.info(f"User {user_id} already has agency {existing.id}, returning existing")
            return existing

        # Create new agency
        agency = Agency(
            owner_user_id=user_id,
            name=name,
            description=description,
            settings={}
        )

        db.add(agency)
        db.commit()
        db.refresh(agency)

        logger.info(f"Created new agency {agency.id} for user {user_id}")

        return agency


# Singleton instance
whitelabel_service = WhitelabelService()
